File: src/decoders/grades-overview.ts
import type { GradesOverview, SessionHandle } from "~/models";
import { decodeSubjectAverages } from "./subject-averages";
import { decodeGradeValue } from "./grade-value";
import { decodeGrade } from "./grade";

export const decodeGradesOverview = (overview: any, session: SessionHandle): GradesOverview => {
  return {
    grades: overview.listeDevoirs.V.map((grade: any) => decodeGrade(grade, session)),
    subjectsAverages: overview.listeServices.V.map(decodeSubjectAverages),
    classAverage: overview.moyGeneraleClasse && decodeGradeValue(overview.moyGeneraleClasse.V),
    overallAverage: overview.moyGenerale && decodeGradeValue(overview.moyGenerale.V)
  };
};

File: library/src/commonMain/kotlin/decoders/grades-overview.kt
package decoders

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import models.GradesOverview
import models.SessionInformation

fun decodeGradesOverview (overview: JsonObject, sessionInformation: SessionInformation): GradesOverview {
    return GradesOverview(
        grades = overview["listeDevoirs"]!!.jsonObject["V"]!!.jsonArray.map {decodeGrade(it.jsonObject, sessionInformation)},
        subjectsAverages = overview["listeServices"]!!.jsonObject["V"]!!.jsonArray.map { decodeSubjectAverages(it.jsonObject) },
        classAverage = if (overview["moyGeneraleClasse"] != null) decodeGradeValue(overview["moyGeneraleClasse"]!!.jsonObject["V"]!!.jsonPrimitive.content) else null,
        overallAverage = if (overview["moyGenerale"] != null) decodeGradeValue(overview["moyGenerale"]!!.jsonObject["V"]!!.jsonPrimitive.content) else null
    )
}